packages/@ember/-internals/glimmer/lib/component-managers/custom.ts

```diff
@@ -150,16 +154,86 @@ export default class CustomComponentManager<ComponentInstance>
     const { delegate } = definition;
     const capturedArgs = args.capture();
 
-    const component = delegate.createComponent(
-      definition.ComponentClass.class,
-      capturedArgs.value()
-    );
+    let value;
+    let namedArgsProxy;
+
+    if (EMBER_CUSTOM_COMPONENT_ARG_PROXY) {
+      if (HAS_NATIVE_PROXY) {
+        let handler: ProxyHandler<{}> = {
+          get(_target, prop) {
+            assert('args can only be strings', typeof prop === 'string');
+
+            let tracker = getCurrentTracker();
+            let ref = capturedArgs.named.get(prop as string);
+
+            if (tracker) {
+              tracker.add(ref.tag);
+            }
+
+            return ref.value();
+          },
+        };
+
+        if (DEBUG) {
+          handler.set = function(_target, prop) {
+            assert(
+              `You attempted to set ${String(
+                prop
```

Comment:
Might be nice to include the component instance in this message (will print a nice `<app-name/components/foo#1234>` to let you know which component) in this message.